### PR TITLE
implement execution janitor

### PR DESCRIPTION
**Pull request by luluz66**
- Refactor janitor struct to re-use the code to clean up executions and
invocations.
- implement the logic to delete expired Executions and InvocationExecutions from
MySQL

**Related issues**:
https://github.com/buildbuddy-io/buildbuddy-internal/issues/2172

### CODE REVIEW

File: server/janitor/janitor.go [hunk 1]
```diff
@@ -1,79 +1,180 @@
...
 	"github.com/buildbuddy-io/buildbuddy/server/tables"
+	"github.com/buildbuddy-io/buildbuddy/server/util/db"
 	"github.com/buildbuddy-io/buildbuddy/server/util/log"
 )
 
 var (
+	// Flags shared by both invocation and execution janitor.
+	logDeletionErrors = flag.Bool("log_deletion_errors", false, "If true; log errors when ttl-deleting expired data")
+
+	// Flags for Invocaction Janitor.
```
Comment (vadimberezniker) [inline server/janitor/janitor.go:19]: Invocation

File: server/janitor/janitor.go [hunk 2]
```diff
@@ -1,79 +1,180 @@
...
 	ttlSeconds = flag.Int("storage.ttl_seconds", 0, "The time, in seconds, to keep invocations before deletion. 0 disables invocation deletion.")
 
-	cleanupBatchSize  = flag.Int("storage.cleanup_batch_size", 10, "How many invocations to delete in each janitor cleanup task")
-	cleanupInterval   = flag.Duration("cleanup_interval", 10*60*time.Second, "How often the janitor cleanup tasks will run")
-	cleanupWorkers    = flag.Int("cleanup_workers", 1, "How many cleanup tasks to run")
-	logDeletionErrors = flag.Bool("log_deletion_errors", false, "If true; log errors when ttl-deleting expired data")
+	cleanupBatchSize = flag.Int("storage.cleanup_batch_size", 10, "How many invocations to delete in each janitor cleanup task")
```
Comment (vadimberezniker) [inline server/janitor/janitor.go:22]: Can you add an 'invocation' prefix to the flag variables? We can leave the flag names as-is.

File: server/janitor/janitor.go [hunk 3]
```diff
@@ -1,79 +1,180 @@
...
+	cleanupInterval  = flag.Duration("cleanup_interval", 10*60*time.Second, "How often the janitor cleanup tasks will run")
+	cleanupWorkers   = flag.Int("cleanup_workers", 1, "How many cleanup tasks to run")
+
+	// Flags for Invocaction Janitor.
+	executionTTLSeconds = flag.Int("storage.execution.ttl_seconds", 0, "The time, in seconds, to keep invocations before deletion. 0 disables invocation deletion.")
```
Comment (vadimberezniker) [inline server/janitor/janitor.go:27]: Ideally this should be a Duration flag, but I can see why you might want to keep it as-is for consistency.
Maybe we can fix this by introducing a new Duration flag for invocations and deprecating the old one.

**Review by vadimberezniker — APPROVED** [review]

File: server/janitor/janitor.go [hunk 4]
```diff
@@ -1,79 +1,180 @@
...
 
-	cleanupBatchSize  = flag.Int("storage.cleanup_batch_size", 10, "How many invocations to delete in each janitor cleanup task")
-	cleanupInterval   = flag.Duration("cleanup_interval", 10*60*time.Second, "How often the janitor cleanup tasks will run")
-	cleanupWorkers    = flag.Int("cleanup_workers", 1, "How many cleanup tasks to run")
-	logDeletionErrors = flag.Bool("log_deletion_errors", false, "If true; log errors when ttl-deleting expired data")
+	cleanupBatchSize = flag.Int("storage.cleanup_batch_size", 10, "How many invocations to delete in each janitor cleanup task")
+	cleanupInterval  = flag.Duration("cleanup_interval", 10*60*time.Second, "How often the janitor cleanup tasks will run")
+	cleanupWorkers   = flag.Int("cleanup_workers", 1, "How many cleanup tasks to run")
+
+	// Flags for Invocaction Janitor.
```
Comment (vadimberezniker) [inline server/janitor/janitor.go:26]: Execution

Comment (luluz66) [reply server/janitor/janitor.go]: Done.

File: server/janitor/janitor.go (hunk 3 above)
Comment (luluz66) [reply server/janitor/janitor.go]: I changed the one for execution into a duration and leave invocation as it is.

File: server/janitor/janitor.go (hunk 2 above)
Comment (luluz66) [reply server/janitor/janitor.go]: Done

File: server/janitor/janitor.go (hunk 1 above)
Comment (luluz66) [reply server/janitor/janitor.go]: Done

**Pull request merged**